### PR TITLE
Blog App | Processing Data 🎯

### PR DESCRIPTION
**Pull request by Rysth**
# 🏅 Features Implemented

List of things that were implemented based on the requirements:

## 📝 General requirements

- ✅ No Linters errors.
- ✅ Used GitFlow correctly.
- ✅ Created a professional `README` file.

## 📝 Project requirements

- ✅ Created model User with method `recent_post`.
- ✅ Created model Post with method `update_counter` & `recent_comments`.
- ✅ Created model Comment with method `update_comments_counter`.
- ✅ Created model Likes with method `update_likes_counter`.
- ✅ Added associations between User, Post, Like, and Comment.

### User
```ruby

   CREATE | Queries to create a User

     first_user = User.create(name: 'Tom', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Mexico.')
     second_user = User.create(name: 'Lilly', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Poland.')

   READ | Queries to count all the User

     User.count()

   UPDATE | Queries to update all the User

     User.all.update_all(name: 'John')

   DELETE | Queries to delete a User by his name and from a specific id.

     User.where(name: 'John').destroy_all
``` 

### Post

```ruby
   CREATE | Queries to create Post

     first_post = Post.create(author_id: first_user.id, title: 'Hello', text: 'This is my first post')
     second_post = Post.create(author_id: first_user.id, title: 'olleH', text: 'This is my second post')
     third_post = Post.create(author_id: first_user.id, title: 'World', text: 'This is my third post')
     fourth_post = Post.create(author_id: first_user.id, title: 'Games', text: 'This is my fourth post')

   READ | Queries to get all the Post

     Post.all

   UPDATE | Queries to update a Post from a specific id.

     Post.find(first_post.id).update(title: 'Videogames')

   DELETE | Queries to delete a Post from a specific id.

     Post.find(first_post.id).destroy
```

### Comments
```ruby
   CREATE | Queries to create Comments

     Comment.create(post: first_post, user: second_user, text: 'Hi Tom!' )
     Comment.create(post: first_post, user: second_user, text: 'Hi John!' )
     Comment.create(post: first_post, user: second_user, text: 'Hi John 2!' )
     Comment.create(post: first_post, user: second_user, text: 'Hi Palacios!' )
     Comment.create(post: first_post, user: second_user, text: 'Hi Rysth!' )
     Comment.create(post: first_post, user: second_user, text: 'Hi Microverse!' )

   READ | Queries to get all the Comments from a specific Post.

     Comment.where(post_id: first_post.id)

   UPDATE | Queries to update a Comment from a specific User.

     Comment.find_by(user_id: second_user.id).update(text: 'Hello World!')

   DELETE | Queries to delete all the Comments from a specific User and Post.

     Comment.where(user_id: second_user.id, post_id: first_post.id).destroy_all
```

[_John Palacios_](https://www.linkedin.com/in/john-palacios-rysthcraft/)
_RysthCraft_

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -1,3 +1,13 @@
...
+  has_many :comments, dependent: :destroy
+  has_many :likes, dependent: :destroy
+
+  def update_counter
+    user.update(posts_counter: user.posts.count)
+  end
+
+  def recent_comments
+    comments.order(created_at: :desc).limit(5)
+  end
```
Comment (Abenezer-Tilahun) [inline app/models/post.rb:12]: - [ Optional ] Kindly I recommend you to use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - after_save to call the` #update_counter` method every time a new post has been saved to database, then make the `update_counter` method private.

 ```
def recent_comments
    comments.order(created_at: :desc).limit(5)
  end

after_save :update_counter

  private

    def update_counter
     user.update(posts_counter: user.posts.count)
    end
```

File: app/models/comment.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class Comment < ApplicationRecord
+  belongs_to :user
+  belongs_to :post
+
+  def update_comments_counter
+    post.update(comments_counter: post.comments.count)
+  end
```
Comment (Abenezer-Tilahun) [inline app/models/comment.rb:7]: - [ Optional ] Kindly I recommend you to use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - after_save to call the #`update_comments_counter `method every time a new post has been saved to database, then make the `update_comments_counter` method private.

 ```
after_save : update_comment_counter

  private

  def update_comments_counter
    post.update(comments_counter: post.comments.count)
  end
```

File: app/models/like.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class Like < ApplicationRecord
+  belongs_to :user
+  belongs_to :post
+
+  def update_likes_counter
+    post.update(likes_counter: post.likes.count)
+  end
```
Comment (Abenezer-Tilahun) [inline app/models/like.rb:7]: - [ Optional ] Kindly I recommend you to use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - after_save to call the #`update_likes_counter `method every time a new post has been saved to database, then make the `update_likes_counter` method private.

 ```
after_save : update_likes_counter

  private

  def update_likes_counter
    post.update(likes_counter: post.likes.count)
  end
```

**Review by Abenezer-Tilahun — APPROVED** [review]
Hi @Rysth 
 
Your project is complete  👍 💯   There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 👏🏻

✔️ No linter errors.
✔️ You used correct Git-Flow.
✔️ Professional README  
✔️ Good Commit Messages.
✔️ Good job for Implemented  Blog App - processing data in models.
## Optional suggestions 
 
_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

**Comment by Rysth** [comment]
Thanks @Abenezer-Tilahun

**Pull request merged**